pkg/sql/rowenc/index_fetch.go

```diff
@@ -103,10 +101,6 @@ func InitIndexFetchSpec(
 	}
 	for i := range s.KeyAndSuffixColumns {
 		col := indexCols[i]
-		if !col.Public() {
-			// Key columns must be public.
-			return fmt.Errorf("column %q (%d) is not public", col.GetName(), col.GetID())
-		}
```

Comment:
Sorry it took me a while to look at this @RaduBerinde .

I'm really not sure how legit removing this check is. Any column which is not public is, by definition, not readable, and it seems to me that this function might be used to perform reads.

If that's not exclusively the case, and if UPSERTs are not considered to be reads, and this function is called to perform a write, then this column state check can be relaxed from "readable" to "writable".

Reply:
The nuance here is that if an UPSERT fails because of a conflict on a non-public column, that *particular* conflicting row value is in a way readable (because it is influencing a user-visible outcome, in this case an error). It's really only a matter of whether we include it in the error message or not.